modules/@apostrophecms/module/index.js

```diff
@@ -446,6 +446,32 @@ module.exports = {
         );
       },
 
+      setCacheControl(req, maxAge) {
+        if (typeof maxAge !== 'number') {
+          self.apos.util.warnDev(`"maxAge" property must be defined as a number in the "${self.__meta.name}" module's cache options"`);
+          return;
+        }
+
+        // A cookie in session doesn't mean we can't cache, nor an empty flash or passport object.
+        // Other session properties must be assumed to be specific to the user, with a possible
+        // impact on the response, and thus mean this request must not be cached.
+        // Same rule as in [express-cache-on-demand](https://github.com/apostrophecms/express-cache-on-demand/blob/master/index.js#L102)
+        const isSessionClearForCaching = Object.entries(req.session).every(([ key, val ]) =>
+          key === 'cookie' || (
+            (key === 'flash' || key === 'passport') && _.isEmpty(val)
+          )
+        );
+        const isSafeToCache = !req.user && isSessionClearForCaching;
+        const cacheControlValue = isSafeToCache ? `max-age=${maxAge}` : 'no-store';
+
+        // When serving a page for a connected user, `req.res.header` gets undefined sometimes, somehow.
+        // The Cache-Control value will still be set to "no-cache" (equivalent to "no-store")
+        // by the admin-bar module, so the result will be the same.
```

Comment:
@boutell any idea?
If it's a problem, I can investigate further but I fear that's out of the scope of this PR.
WDYT?

Reply:
My guess is you're running into a simulated `req` object created by `self.apos.task.getReq`. If you look at `getReq` in the task module, the simulated req object is not quite good enough:

```
          res: {
            redirect(url) {
              req.res.redirectedTo = url;
            }
          },
```

There is no header method on the simulated "res".

Your if statement works, but it would be better to provide a dummy "header" method on the simulated "res" object so that it is a little better at simulating a real Express req/res pair. That method does not have to actually do anything, or it could just store the header into an `req.res.headers` object, which would open the door to using it for unit testing. Up to you.